### PR TITLE
many: seal a fallback object to the recovery boot chain

### PR DESCRIPTION
**Pull request by cmatsuoka**
Seal (and reseal) an object with the ubuntu-data and ubuntu-save encryption keys to the recovery boot chains to provide a fallback access that's not resealed as often in case something happens to the regular run object.

### CODE REVIEW

File: boot/bootchain.go [hunk 1]
```diff
@@ -289,8 +289,10 @@ func bootAssetsToLoadChains(assets []bootAsset, kernelBootFile bootloader.BootFi
 // that we do not store the arrays directly as JSON and we can add
 // other information
 type predictableBootChainsWrapperForStorage struct {
-	ResealCount int                   `json:"reseal-count"`
-	BootChains  predictableBootChains `json:"boot-chains"`
+	ResealCount         int                   `json:"reseal-count"`
+	RecoveryResealCount int                   `json:"recovery-reseal-count"`
+	BootChains          predictableBootChains `json:"boot-chains"`
+	RecoveryBootChains  predictableBootChains `json:"recovery-boot-chains"`
```
Comment (pedronis) [inline boot/bootchain.go:295]: I didn't think of this originally, but probably best is two have two separate files maybe?

File: boot/seal.go [hunk 2]
```diff
@@ -95,40 +98,101 @@ func sealKeyToModeenv(key secboot.EncryptionKey, model *asserts.Model, modeenv *
...
+	// also seal the keys to the recovery bootchains as a fallback
+	modelParams, err := sealKeyModelParams(pbc, roleToBlName)
+	if err != nil {
+		return fmt.Errorf("cannot prepare for fallback key sealing: %v", err)
+	}
+	sealKeyParams := &secboot.SealKeyParams{
+		ModelParams:            modelParams,
+		TPMPolicyAuthKey:       authKey,
+		TPMPolicyAuthKeyFile:   filepath.Join(InstallHostFDEDataDir, "tpm-policy-auth-key"),
+		TPMLockoutAuthFile:     filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
```
Comment (pedronis) [inline boot/seal.go:175]: the fact that we have the files again seems wrong to me, see my comment in secboot

File: boot/seal.go [hunk 3]
```diff
@@ -211,24 +275,87 @@ func resealKeyToModeenv(rootdir string, model *asserts.Model, modeenv *Modeenv,
...
+	if needed {
+		rpbcJSON, _ := json.Marshal(rpbc)
+		logger.Debugf("resealing (%d) to recovery boot chains: %s", nextCount, rpbcJSON)
+
+		roleToBlName := map[bootloader.Role]string{
+			bootloader.RoleRecovery: rbl.Name(),
+			bootloader.RoleRunMode:  bl.Name(),
+		}
+
+		if err := resealRunObjectKeys(pbc, authKeyFile, roleToBlName); err != nil {
```
Comment (pedronis) [inline boot/seal.go:301]: this looks wrong? shouldn't it say fallback?

File: boot/seal.go [hunk 4]
```diff
@@ -211,24 +275,87 @@ func resealKeyToModeenv(rootdir string, model *asserts.Model, modeenv *Modeenv,
...
+	// get model parameters from bootchains
+	modelParams, err := sealKeyModelParams(pbc, roleToBlName)
+	if err != nil {
+		return fmt.Errorf("cannot prepare for fallback key resealing: %v", err)
+	}
+
+	// list all the key files to reseal
+	keyFiles := []string{
+		filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.recovery.sealed-key"),
+		filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.recovery.sealed-key"),
```
Comment (pedronis) [inline boot/seal.go:349]: twice data?

File: boot/seal.go [hunk 5]
```diff
@@ -407,3 +534,19 @@ func isResealNeeded(pbc predictableBootChains, rootdir string, expectReseal bool
 	}
 	return true, c + 1, nil
 }
+
+func isFallbackResealNeeded(pbc predictableBootChains, rootdir string, expectReseal bool) (ok bool, nextCount int, err error) {
```
Comment (pedronis) [inline boot/seal.go:538]: this can probably be one helper somehow  called twice with the right arguments, instead of two full separate helpers

File: boot/seal.go (hunk 2 above)
Comment (pedronis) [reply boot/seal.go:175]: ah, if TPMProvision is fail TPMLockoutAuthFile is not used afaict... see also my comment in secboot

File: secboot/secboot.go [hunk 6]
```diff
@@ -58,19 +73,23 @@ type SealKeyModelParams struct {
 type SealKeyParams struct {
 	// The parameters we're sealing the key to
 	ModelParams []*SealKeyModelParams
-	// The path to store the sealed key file
-	KeyFile string
+	// The authorization policy update key file (only relevant for TPM)
+	TPMPolicyAuthKey *ecdsa.PrivateKey
 	// The path to the authorization policy update key file (only relevant for TPM)
 	TPMPolicyAuthKeyFile string
 	// The path to the lockout authorization file (only relevant for TPM)
```
Comment (pedronis) [inline secboot/secboot.go:80]: this should also say and if TPMProvision is true

File: secboot/secboot.go [hunk 7]
```diff
@@ -58,19 +73,23 @@ type SealKeyModelParams struct {
 type SealKeyParams struct {
 	// The parameters we're sealing the key to
 	ModelParams []*SealKeyModelParams
-	// The path to store the sealed key file
-	KeyFile string
+	// The authorization policy update key file (only relevant for TPM)
+	TPMPolicyAuthKey *ecdsa.PrivateKey
 	// The path to the authorization policy update key file (only relevant for TPM)
```
Comment (pedronis) [inline secboot/secboot.go:78]: this should be optionally used if TPMPolicyAuthKey is set

File: secboot/secboot_tpm.go [hunk 8]
```diff
@@ -377,18 +374,29 @@ func SealKey(key EncryptionKey, params *SealKeyParams) error {
...
+			Key:  keys[i].Key[:],
+			Path: keys[i].KeyFile,
+		})
 	}
 
-	authKey, err := sbSealKeyToTPM(tpm, key[:], params.KeyFile, &creationParams)
+	authKey, err := sbSealKeyToTPMMultiple(tpm, sbKeys, &creationParams)
 	if err != nil {
 		return err
 	}
```
Comment (pedronis) [inline secboot/secboot_tpm.go:402]: the write to the fail after this should be optional if AuthKey was passed in and the file name is empty

**Review by pedronis — COMMENTED** [review]
did a first pass,  one main consideration is whether we want to have 2 files instead of just boot-chains

File: boot/bootchain.go (hunk 1 above)
Comment (cmatsuoka) [reply boot/bootchain.go:295]: That sounds like a good idea, I'll explore it. It also should reduce some duplication in data access functions.

File: boot/bootchain.go [hunk 9]
```diff
@@ -309,7 +311,23 @@ func readBootChains(path string) (pbc predictableBootChains, resealCount int, er
...
+		if os.IsNotExist(err) {
+			return nil, 0, nil
+		}
+		return nil, 0, fmt.Errorf("cannot open existing boot chains data file: %v", err)
+	}
+	defer inf.Close()
+	var wrapped predictableBootChainsWrapperForStorage
+	if err := json.NewDecoder(inf).Decode(&wrapped); err != nil {
+		return nil, 0, fmt.Errorf("cannot read boot chains data: %v", err)
+	}
```
Comment (bboozzoo) [inline boot/bootchain.go:326]: this could be a shared helper

File: boot/seal.go (hunk 3 above)
Comment (cmatsuoka) [reply boot/seal.go:301]: It's completely wrong. Fixing it.

File: boot/seal.go (hunk 4 above)
Comment (cmatsuoka) [reply boot/seal.go:349]: Also wrong, late night coding while tired is showing up. Fixing. (To be fair I didn't have time to do a re-check pass on the code before opening the draft PR so I'm not so surprised with typos or errors like this)

File: secboot/secboot_tpm.go [hunk 10]
```diff
@@ -354,10 +351,10 @@ func unlockEncryptedPartitionWithKey(name, device string, key []byte) error {
 	return err
 }
 
-// SealKey provisions the TPM and seals a partition encryption key according to the
+// SealKey provisions the TPM and seals the encryption keys according to the
 // specified parameters. If the TPM is already provisioned, or a sealed key already
 // exists, SealKey will fail and return an error.
-func SealKey(key EncryptionKey, params *SealKeyParams) error {
+func SealKey(keys []SealKeyRequest, params *SealKeyParams) error {
```
Comment (bboozzoo) [inline secboot/secboot_tpm.go:357]: SealKeys?

File: boot/seal.go [hunk 11]
```diff
@@ -211,24 +275,87 @@ func resealKeyToModeenv(rootdir string, model *asserts.Model, modeenv *Modeenv,
...
 	}
-	logger.Debugf("resealing (%d) succeeded", nextCount)
 
-	bootChainsPath := bootChainsFileUnder(rootdir)
-	if err := writeBootChains(pbc, bootChainsPath, nextCount); err != nil {
-		return err
+	return nil
+}
+
+func resealFallbackObjectKeys(pbc predictableBootChains, authKeyFile string, roleToBlName map[bootloader.Role]string) error {
```
Comment (bboozzoo) [inline boot/seal.go:339]: This does not appear to be called anywhere?

File: boot/seal.go [hunk 12]
```diff
@@ -95,40 +98,101 @@ func sealKeyToModeenv(key secboot.EncryptionKey, model *asserts.Model, modeenv *
...
-	sealKeyParams := &secboot.SealKeyParams{
-		ModelParams:          modelParams,
-		KeyFile:              filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
-		TPMPolicyAuthKeyFile: filepath.Join(InstallHostFDEDataDir, "tpm-policy-auth-key"),
-		TPMLockoutAuthFile:   filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
+
+	// the boot chains we seal the fallback object to
+	rpbc := toPredictableBootChains(recoveryBootChains)
+
+	authKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
```
Comment (bboozzoo) [inline boot/seal.go:111]: looks like secboot.SealKeyToTPM() generates and returns the, would it make sense to use it rather than generate one ourselves?

File: boot/bootchain.go (hunk 9 above)
Comment (cmatsuoka) [reply boot/bootchain.go:326]: Yes, especially if we go with a separate file for the fallback chains.

File: secboot/secboot_tpm.go (hunk 10 above)
Comment (cmatsuoka) [reply secboot/secboot_tpm.go:357]: SealKeys is a better name, will replace. I just copied the name used in snapcore/secboot to get a quicker start, but I don't think we should keep that name.

File: boot/seal.go (hunk 11 above)
Comment (cmatsuoka) [reply boot/seal.go:339]: There was a bug in the code where the run object reseal was being called instead. Fixing.

File: boot/seal.go (hunk 12 above)
Comment (cmatsuoka) [reply boot/seal.go:111]: Yes, I think we could get it from the first seal operation and reuse it in the second.

Comment (cmatsuoka) [reply boot/seal.go:111]: In fact we can't because the auth key returned by the seal operation is a derivative of the supplied auth key (the elliptic curve private key). Maybe there's a way to get back to the original key from the returned data and if this is the case we can refactor later.

File: boot/bootchain.go (hunk 9 above)
Comment (cmatsuoka) [reply boot/bootchain.go]: Using a separate file for recovery bootchains so all helpers can be used without duplication.

File: boot/seal.go (hunk 5 above)
Comment (cmatsuoka) [reply boot/seal.go]: Using a separate file for recovery boot chains allows us to use the same helper for both cases (but passing the full file path instead of just the rootdir).

File: boot/seal.go (hunk 2 above)
Comment (cmatsuoka) [reply boot/seal.go]: Changed to not pass the auth key filename in the second call (and in this case the file won't be written again).

File: boot/seal.go [hunk 13]
```diff
@@ -127,6 +135,71 @@ func sealKeyToModeenv(key secboot.EncryptionKey, model *asserts.Model, modeenv *
...
+func sealFallbackObjectKeys(key, saveKey secboot.EncryptionKey, pbc predictableBootChains, authKey *ecdsa.PrivateKey, roleToBlName map[bootloader.Role]string) error {
+	// also seal the keys to the recovery bootchains as a fallback
+	modelParams, err := sealKeyModelParams(pbc, roleToBlName)
+	if err != nil {
+		return fmt.Errorf("cannot prepare for fallback key sealing: %v", err)
+	}
+	sealKeyParams := &secboot.SealKeyParams{
+		ModelParams:            modelParams,
+		TPMPolicyAuthKey:       authKey,
+		TPMLockoutAuthFile:     filepath.Join(InstallHostFDESaveDir, "tpm-lockout-auth"),
```
Comment (pedronis) [inline boot/seal.go:185]: this doesn't need to be set I think given that TPMProvision is false

File: boot/seal.go [hunk 14]
```diff
@@ -212,24 +285,87 @@ func resealKeyToModeenv(rootdir string, model *asserts.Model, modeenv *Modeenv,
...
+		logger.Debugf("fallback resealing (%d) succeeded", nextFallbackCount)
+	}
+
+	bootChainsPath := bootChainsFileUnder(rootdir)
+	if err := writeBootChains(pbc, bootChainsPath, nextCount); err != nil {
+		return err
+	}
+
+	recoveryBootChainsPath := recoveryBootChainsFileUnder(rootdir)
+	if err := writeBootChains(rpbc, recoveryBootChainsPath, nextFallbackCount); err != nil {
```
Comment (pedronis) [inline boot/seal.go:318]: this write should go under the `if needed ...{`, no need to write the file if we didn't detect that reseal was needed, we do the same for the main one

**Review by pedronis — COMMENTED** [review]
thanks for the changes, couple of comments

File: boot/seal.go [hunk 15]
```diff
@@ -95,27 +102,28 @@ func sealKeyToModeenv(key secboot.EncryptionKey, model *asserts.Model, modeenv *
...
-	sealKeyParams := &secboot.SealKeyParams{
-		ModelParams:          modelParams,
-		KeyFile:              filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
-		TPMPolicyAuthKeyFile: filepath.Join(InstallHostFDESaveDir, "tpm-policy-auth-key"),
-		TPMLockoutAuthFile:   filepath.Join(InstallHostFDESaveDir, "tpm-lockout-auth"),
+
+	// the boot chains we seal the fallback object to
+	rpbc := toPredictableBootChains(recoveryBootChains)
+
+	authKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
```
Comment (mvo5) [inline boot/seal.go:116]: I was a bit confused initially where this authKey is saved. Maybe a small comment that the "seal*ObjectKeys()" will save the key to disk?

File: secboot/secboot_tpm.go (hunk 10 above)
Comment (mvo5) [reply secboot/secboot_tpm.go:357]: Yeah, had the same question

**Review by mvo5 — COMMENTED** [review]
This looks good, I would love to have @chrisccoulson checking this as well (the secboot/encryption part not so much the other bits :)

File: secboot/secboot_tpm.go (hunk 10 above)
Comment (bboozzoo) [reply secboot/secboot_tpm.go:357]: Renamed to SealKeys & ResealKeys

**Review by bboozzoo — APPROVED** [review]

File: boot/makebootable_test.go [hunk 16]
```diff
@@ -412,9 +414,19 @@ version: 5.0
...
+		switch sealKeyCalls {
+		case 1:
+			c.Check(keys, HasLen, 1)
+			c.Check(keys[0].Key, DeepEquals, myKey)
+		case 2:
+			c.Check(keys, HasLen, 2)
+			c.Check(keys[0].Key, DeepEquals, myKey)
+			c.Check(keys[1].Key, DeepEquals, myKey2)
+		case 3:
+			c.Error("secboot.SealKey shouldn't be called a third time")
```
Comment (anonymouse64) [inline boot/makebootable_test.go:428]: see above

File: boot/makebootable_test.go [hunk 17]
```diff
@@ -429,14 +441,25 @@ version: 5.0
...
+				"snapd_recovery_mode=recover snapd_recovery_system=20191216 console=ttyS0 console=tty1 panic=-1",
+				"snapd_recovery_mode=run console=ttyS0 console=tty1 panic=-1",
+			})
+		case 2:
+			c.Assert(params.ModelParams[0].EFILoadChains, DeepEquals, []*secboot.LoadChain{
+				secboot.NewLoadChain(shim, secboot.NewLoadChain(grub, secboot.NewLoadChain(kernel))),
+			})
+			c.Assert(params.ModelParams[0].KernelCmdlines, DeepEquals, []string{
+				"snapd_recovery_mode=recover snapd_recovery_system=20191216 console=ttyS0 console=tty1 panic=-1",
+			})
```
Comment (anonymouse64) [inline boot/makebootable_test.go:460]: this should also have a `default` with the above message/error returned

File: boot/makebootable_test.go [hunk 18]
```diff
@@ -729,9 +754,19 @@ version: 5.0
...
+		switch sealKeyCalls {
+		case 1:
+			c.Check(keys, HasLen, 1)
+			c.Check(keys[0].Key, DeepEquals, myKey)
+		case 2:
+			c.Check(keys, HasLen, 2)
+			c.Check(keys[0].Key, DeepEquals, myKey)
+			c.Check(keys[1].Key, DeepEquals, myKey2)
+		default:
+			c.Error("secboot.SealKey shouldn't be called a third time")
```
Comment (anonymouse64) [inline boot/makebootable_test.go:768]: same here

File: boot/seal_test.go [hunk 19]
```diff
@@ -122,10 +124,22 @@ func (s *sealSuite) TestSealKeyToModeenv(c *C) {
...
+				// the run object seals only the ubuntu-data key
+				dataKeyFile := filepath.Join(rootdir, "/run/mnt/ubuntu-seed/device/fde/ubuntu-data.sealed-key")
+				c.Check(keys, DeepEquals, []secboot.SealKeyRequest{{Key: myKey, KeyFile: dataKeyFile}})
+			case 2:
+				// the fallback object seals the ubuntu-data and the ubuntu-save keys
+				dataKeyFile := filepath.Join(rootdir, "/run/mnt/ubuntu-seed/device/fde/ubuntu-data.recovery.sealed-key")
+				saveKeyFile := filepath.Join(rootdir, "/run/mnt/ubuntu-seed/device/fde/ubuntu-save.recovery.sealed-key")
+				c.Check(keys, DeepEquals, []secboot.SealKeyRequest{{Key: myKey, KeyFile: dataKeyFile}, {Key: myKey2, KeyFile: saveKeyFile}})
+			default:
+				c.Error("secboot.SealKey shouldn't be called a third time")
```
Comment (anonymouse64) [inline boot/seal_test.go:141]: same here

File: boot/seal_test.go [hunk 20]
```diff
@@ -138,27 +152,43 @@ func (s *sealSuite) TestSealKeyToModeenv(c *C) {
...
+				})
+			case 2:
+				c.Assert(params.ModelParams[0].EFILoadChains, DeepEquals, []*secboot.LoadChain{
+					secboot.NewLoadChain(shim,
+						secboot.NewLoadChain(grub,
+							secboot.NewLoadChain(kernel))),
+				})
+				c.Assert(params.ModelParams[0].KernelCmdlines, DeepEquals, []string{
+					"snapd_recovery_mode=recover snapd_recovery_system=20200825 console=ttyS0 console=tty1 panic=-1",
+				})
```
Comment (anonymouse64) [inline boot/seal_test.go:178]: here too

File: boot/seal_test.go [hunk 21]
```diff
@@ -313,20 +374,30 @@ func (s *sealSuite) TestResealKeyToModeenv(c *C) {
...
+				// load chains
+				c.Assert(params.ModelParams[0].EFILoadChains, HasLen, 6)
+			case 2:
+				c.Assert(params.ModelParams[0].KernelCmdlines, DeepEquals, []string{
+					"snapd_recovery_mode=recover snapd_recovery_system=20200825 console=ttyS0 console=tty1 panic=-1",
+				})
+				// load chains
+				c.Assert(params.ModelParams[0].EFILoadChains, HasLen, 2)
+			default:
+				c.Error("secboot.ResealKey shouldn't be called a third time")
```
Comment (anonymouse64) [inline boot/seal_test.go:399]: same

File: boot/seal_test.go [hunk 22]
```diff
@@ -349,39 +420,42 @@ func (s *sealSuite) TestResealKeyToModeenv(c *C) {
...
+						)),
+					secboot.NewLoadChain(shim2,
+						secboot.NewLoadChain(grub,
+							secboot.NewLoadChain(runGrub,
+								secboot.NewLoadChain(runKernel2)),
+							secboot.NewLoadChain(runGrub2,
+								secboot.NewLoadChain(runKernel2)),
+						)),
+				})
+			}
```
Comment (anonymouse64) [inline boot/seal_test.go:458]: this should have a default too I think

File: boot/seal_test.go [hunk 23]
```diff
@@ -764,49 +842,50 @@ func (s *sealSuite) TestIsResealNeeded(c *C) {
 	err := boot.WriteBootChains(pbc, filepath.Join(dirs.SnapFDEDirUnder(rootdir), "boot-chains"), 2)
 	c.Assert(err, IsNil)
 
-	needed, _, err := boot.IsResealNeeded(pbc, rootdir, false)
+	needed, _, err := boot.IsResealNeeded(pbc, filepath.Join(dirs.SnapFDEDirUnder(rootdir), "boot-chains"), false)
```
Comment (anonymouse64) [inline boot/seal_test.go:845]: we should have similar tests of IsResealNeeded for the recovery-boot-chains file too, shouldn't we (unless I somehow missed this testing somewhere)?

File: secboot/secboot_tpm.go [hunk 24]
```diff
@@ -377,31 +374,44 @@ func SealKey(key EncryptionKey, params *SealKeyParams) error {
 		return err
 	}
 
-	// Provision the TPM as late as possible
-	if err := tpmProvision(tpm, params.TPMLockoutAuthFile); err != nil {
-		return err
+	if params.TPMProvision {
```
Comment (anonymouse64) [inline secboot/secboot_tpm.go:377]: what's the situation where this will be false? I don't see any change where this will be set to false anywhere, is this just to prepare for eventually not having a TPM in this codepath?

**Review by anonymouse64 — APPROVED** [review]
Overall looks good, I have one question about TPMProvision and it's intended usage, and the tests I think should be expanded to explicitly return non-nil error on the 3rd (or more) calls to ResealKey(), as this makes the failed test output more understandable what happened, as today we have multiple codepaths in the tests where a nil error ends up being returned for a case where ResealKey was called too many times, and this could make the entire test pass except for the final message which says "ResealKey shouldn't be called for a third time" which would be confusing to debug I think if it ever happens.

also one clarification on the ordering of resealing to the run vs fallback objects, I think it's not a problem but want to have confirmation

File: boot/makebootable_test.go [hunk 25]
```diff
@@ -429,14 +441,25 @@ version: 5.0
...
+		case 2:
+			c.Assert(params.ModelParams[0].EFILoadChains, DeepEquals, []*secboot.LoadChain{
+				secboot.NewLoadChain(shim, secboot.NewLoadChain(grub, secboot.NewLoadChain(kernel))),
+			})
+			c.Assert(params.ModelParams[0].KernelCmdlines, DeepEquals, []string{
+				"snapd_recovery_mode=recover snapd_recovery_system=20191216 console=ttyS0 console=tty1 panic=-1",
+			})
+		}
+
 		c.Assert(params.ModelParams[0].Model.DisplayName(), Equals, "My Model")
```
Comment (mvo5) [inline boot/makebootable_test.go:463]: This does not check how many times sealKeyCalsl happend

File: boot/seal_test.go (hunk 22 above)
Comment (mvo5) [reply boot/seal_test.go:458]: When I add this it starts to fail it seems :(

Comment (bboozzoo) [reply boot/seal_test.go:460]: It's because the recovery boot chains are common for both keys and are checked in line 408 https://github.com/snapcore/snapd/pull/9577/files/74e5eaef07bc091e158d0c2b961fb406f4517847#diff-fda1ca0460d96a5f1f24e0fef2c6e29bdd73e9a28c4ddee02a01c14c3ced24d8R408

Now, to double check, the run mode data key is supposed to sealed to both recovery and run mode load chains, right?
While the fallback keys are sealed only to the recovery load chains?

Comment (anonymouse64) [reply boot/seal_test.go:460]: @bboozzoo 

> Now, to double check, the run mode data key is supposed to sealed to both recovery and run mode load chains, right?

yes because the first key we try to use to unseal / decrypt ubuntu-data in recover mode is the one on ubuntu-boot, if that fails then we fallback to the one on ubuntu-seed (which is the fallback key aiui)

> While the fallback keys are sealed only to the recovery load chains?

I think so, that's my understanding at least, is that we only intend to use the fallback keys from recovery (and at least that's what the uc20 save TODO's doc says)

File: boot/makebootable_test.go [hunk 26]
```diff
@@ -728,10 +755,20 @@ version: 5.0
 	defer restore()
 
 	// set mock key sealing
-	sealKeyCalls := 0
-	restore = boot.MockSecbootSealKey(func(key secboot.EncryptionKey, params *secboot.SealKeyParams) error {
-		sealKeyCalls++
-		c.Check(key, DeepEquals, myKey)
+	sealKeysCalls := 0
```
Comment (mvo5) [inline boot/makebootable_test.go:758]: This var is not actually used in the test. The test also tests the error case and will error after the first seal call so the switch below with "case 2" is never hit. I guess this should just be split into erroring once for "sealKeysCall" calls 1 and then calls 2.

**Comment by pedronis** [comment]
On Mon, 2 Nov 2020 at 17:53, Ian Johnson <notifications@github.com> wrote:

> *@anonymouse64* commented on this pull request.
> ------------------------------
>
> In boot/seal_test.go
> <https://github.com/snapcore/snapd/pull/9577#discussion_r516112428>:
>
> > +					secboot.NewLoadChain(shim,
> +						secboot.NewLoadChain(grub,
> +							secboot.NewLoadChain(runGrub,
> +								secboot.NewLoadChain(runKernel2)),
> +							secboot.NewLoadChain(runGrub2,
> +								secboot.NewLoadChain(runKernel2)),
> +						)),
> +					secboot.NewLoadChain(shim2,
> +						secboot.NewLoadChain(grub,
> +							secboot.NewLoadChain(runGrub,
> +								secboot.NewLoadChain(runKernel2)),
> +							secboot.NewLoadChain(runGrub2,
> +								secboot.NewLoadChain(runKernel2)),
> +						)),
> +				})
> +			}
>
> @bboozzoo <https://github.com/bboozzoo>
>
> Now, to double check, the run mode data key is supposed to sealed to both
> recovery and run mode load chains, right?
>
> yes because the first key we try to use to unseal / decrypt ubuntu-data in
> recover mode is the one on ubuntu-boot, if that fails then we fallback to
> the one on ubuntu-seed (which is the fallback key aiui)
>
> While the fallback keys are sealed only to the recovery load chains?
>
> I think so, that's my understanding at least, is that we only intend to
> use the fallback keys from recovery (and at least that's what the uc20 save
> TODO's doc says)
>

yes, otherwise they would need to cover the run boot chains and then that
defeats their purpose because they would need to be resealed all the time

the run ones instead can cover all the paths, we need to reseal them often
anyway, but that way they can still save us in case we break the fallback
ones

belt-and-suspenders

File: boot/seal_test.go (hunk 22 above)
Comment (anonymouse64) [reply boot/seal_test.go:460]: > yes because the first key we try to use to unseal / decrypt ubuntu-data in recover mode is the one on ubuntu-boot, if that fails then we fallback to the one on ubuntu-seed (which is the fallback key aiui)

just a note to myself - all keys exist on ubuntu-seed, we don't put any keys on ubuntu-boot currently
